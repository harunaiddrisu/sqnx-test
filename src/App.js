import React, { Component } from 'react'
import axios from 'axios'
import {
  appStyles,
  gridContainerStyles,
  firstRowStyles,
  secondRowStyles,
  thirdRowStyles,
} from './styles.js'

const API_URL = ' http://demo9164717.mockable.io'

export class App extends Component {
  _isMounted = false
  state = { apiData: this.props.apiData || [] }

  componentDidMount() {
    this._isMounted = true
    axios
      .get(API_URL)
      .then(({ data }) => {
        // handle success
        // console.log(data)
        if (this._isMounted) {
          this.setState(() => ({ apiData: data }))
        }
      })
      .catch(error => {
        // handle error
        console.log(error)
      })
  }

  componentWillUnmount() {
    this._isMounted = false
  }

  render() {
    const { apiData = [] } = this.state
    return <Grid apiData={apiData} />
  }
}

export const Grid = ({ apiData }) => {
  return (
    <div {...appStyles}>
      <div {...gridContainerStyles}>
        <section {...firstRowStyles}>
          {apiData.length !== 0 &&
            apiData
              .slice(0, 5)
              .map((item, index) => <GridItem key={item.id} item={item} />)}
        </section>
        <section {...secondRowStyles}>
          {apiData.length !== 0 &&
            apiData
              .slice(5, 9)
              .map((item, index) => <GridItem key={item.id} item={item} />)}
        </section>
        <section {...thirdRowStyles}>
          {apiData.length !== 0 &&
            apiData
              .slice(9, 15)
              .map((item, index) => <GridItem key={item.id} item={item} />)}
        </section>
      </div>
    </div>
  )
}

export const GridItem = ({ item }) => (
  <div>
    <figure>
      <img src={item.artwork} alt={item.title} />
      <figcaption>
        <h3>{`${item.id}. ${item.title}`}</h3>
        {item.content}
      </figcaption>
    </figure>
  </div>
)

export default App
