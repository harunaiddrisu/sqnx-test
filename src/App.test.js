import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import axios from 'axios'
import { App, Grid, GridItem } from './App'
import fakeApi from './fakeApi.json'

jest.mock('axios')
it('fetches mocked data', () => {
  axios.get.mockImplementation(() => Promise.resolve({ data: fakeApi }))
})
it('renders app without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<App />, div)
  ReactDOM.unmountComponentAtNode(div)
})
it('renders grid correctly', () => {
  const app = renderer.create(<Grid apiData={fakeApi} />).toJSON()
  expect(app).toMatchSnapshot()
})
it('renders grid item correctly', () => {
  const app = renderer.create(<GridItem item={fakeApi[0]} />).toJSON()
  expect(app).toMatchSnapshot()
})
