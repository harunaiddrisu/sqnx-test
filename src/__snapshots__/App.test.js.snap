// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders grid correctly 1`] = `
<div
  data-css-q2vgoc=""
>
  <div
    data-css-12k7nt2=""
  >
    <section
      data-css-1u9xw1i=""
    >
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              1. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              2. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              3. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              4. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              5. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
    </section>
    <section
      data-css-12k81u0=""
    >
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              6. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              7. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              8. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              9. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
    </section>
    <section
      data-css-j174in=""
    >
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              10. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              11. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              12. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              13. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              14. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
      <div>
        <figure>
          <img
            alt="Lorem Ipsum"
            src="https://placehold.it/1920x1080"
          />
          <figcaption>
            <h3>
              15. Lorem Ipsum
            </h3>
            Lorem ipsum dolor sit amet, consectetur...
          </figcaption>
        </figure>
      </div>
    </section>
  </div>
</div>
`;

exports[`renders grid item correctly 1`] = `
<div>
  <figure>
    <img
      alt="Lorem Ipsum"
      src="https://placehold.it/1920x1080"
    />
    <figcaption>
      <h3>
        1. Lorem Ipsum
      </h3>
      Lorem ipsum dolor sit amet, consectetur...
    </figcaption>
  </figure>
</div>
`;
